File: src/role_gate.py
import requests

class RoleGate:
	def __init__(self) -> None:
		self.api = "https://www.rolegate.com"
		self.headers = {
			"user-agent": "okhttp/3.14.9"
		}
		self.token = None
		self.user_id = None

	def register(
			self,
			username: str,
			email: str,
			password: str) -> dict:
		data = {
			"username": username,
			"password": password,
			"password2": password,
			"email": email,
		}
		return requests.post(
			f"{self.api}/account/signup",
			data=data,
			headers=self.headers).json()

	def login(
			self,
			username: str,
			password: str) -> dict:
		data = {
			"username": username,
			"password": password
		}
		response = requests.post(
			f"{self.api}/api/v2/api-token-auth",
			data=data,
			headers=self.headers).json()
		if "token" in response:
			self.token = response["token"]
			self.headers["authorization"] = f"Token {self.token}"
			self.user_id = self.get_current_user()["profile"]["id"]
		return response

	def get_current_user(self) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/current",
			headers=self.headers).json()

	def get_avatars(self) -> dict:
		return requests.get(
			f"{self.api}/api/v2/avatars",
			headers=self.headers).json()

	def get_user_transactions(
			self,
			user_id: int,
			ordering: str = "date",
			page_size: int = 50) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{user_id}/coin-transactions?ordering=-{ordering}&page-size={page_size}",
			headers=self.headers).json()

	def get_user_referring_count(
			self, 
			user_id: int) -> int:
		return requests.get(
			f"{self.api}/api/v2/users/{user_id}/profiles-referring-count",
			headers=self.headers).text

	def get_user_game_alerts(
			self,
			user_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{user_id}/game-alerts",
			headers=self.headers).json()

	def get_user_bookmarks(
			self,
			user_id: int,
			ordering: str = "last_update") -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{user_id}/games?ordering=-{ordering}&bookmarked=True",
			headers=self.headers).json()

	def get_featured_games(self) -> list:
		return requests.get(
			F"{self.api}/api/v2/games/featured-games",
			headers=self.headers).json()

	def get_game_info(self, game_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}",
			headers=self.headers).json()

	def get_game_characters(
			self,
			game_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}/characters",
			headers=self.headers).json()

	def get_game_preview_selection(
			self,
			game_id: int) -> list:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}/lines/preview-selection",
			headers=self.headers).json()

	def get_game_states(
			self,
			game_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}/states",
			headers=self.headers).json()

	def get_game_sheets(
			self,
			game_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}/sheets",
			headers=self.headers).json()

	def get_game_tokens(
			self,
			game_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}/tokens",
			headers=self.headers).json()

	def get_game_dice_sets(
			self,
			game_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games/{game_id}/dice-sets",
			headers=self.headers).json()

	def get_recommended_tags(self) -> list:
		return requests.get(
			f"{self.api}/api/v2/tags/recommended-tags",
			headers=self.headers).json()

	def get_games_by_tag(
			self,
			tag_id: int,
			page_size: int = 10) -> dict:
		return requests.get(
			f"{self.api}/api/v2/tags/{tag_id}/games?page-size={page_size}",
			headers=self.headers).json()

	def get_games(
			self,
			page: int,
			is_hidden: bool = False) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games?hidden={is_hidden}&page={page}",
			headers=self.headers).json()

	def join_game(
			self,
			character_name: str,
			portrait_avatar: str = None,
			color: int = 1,
			is_npc: bool = False) -> dict:
		data = {
			"color": color,
			"background_color": 1,
			"background_rank": "no",
			"background_key": 0,
			"player-id": self.user_id,
			"npc": is_npc,
			"character-name": character_name
		}
		if portrait_avatar:
			data["portrait_avatar"] = portrait_avatar
		return requests.post(
			f"{self.api}/api/v2/games/{game_id}/characters",
			data=data,
			headers=self.headers).json()

	def leave_game(
			self,
			game_id: int) -> int:
		return requests.post(
			f"{self.api}/api/v2/users/{self.user_id}/games/{game_id}/leave",
			headers=self.headers).status_code

	def search_game(
			self,
			query: str,
			page: int = 1,
			is_hidden: bool = False) -> dict:
		return requests.get(
			f"{self.api}/api/v2/games?hidden={is_hidden}&page={page}&search={query}",
			headers=self.headers).json()

	def get_backgrounds(
			self,
			page_size: int = 100) -> dict:
		return requests.get(
			f"{self.api}/api/v2/stage-backgrounds?page-size={page_size}",
			headers=self.headers).json()

	def get_sheets(
			self,
			is_public: bool = True) -> dict:
		return requests.get(
			f"{self.api}/api/v2/sheets?public={is_public}",
			headers=self.headers).json()

	def create_game(
			self,
			name: str,
			description: str,
			rules: str,
			cover: str,
			sheet_id: int,
			public_template_id: int,
			expected_pace: int = 0,
			is_hidden: bool = False,
			allow_spectators: bool = True,
			allow_public_chat: bool = True,
			max_players: int = 4,
			password: str = None,
			tags: list = [],
			dice_sets: list = [1]) -> dict:
		data = {
			"expected_pace": expected_pace,
			"hidden": is_hidden,
			"allow_spectators": allow_spectators,
			"allow_public_chat": allow_public_chat,
			"max_players": max_players,
			"name": name,
			"description": description,
			"rules": rules,
			"password": password,
			"cover": cover,
			"sheet": sheet_id,
			"tags": tags,
			"dice_sets": dice_sets,
			"public_template": public_template_id
		}
		return requests.post(
			f"{self.api}/api/v2/games",
			data=data,
			headers=self.headers).json()


	def get_sheet_info(self, sheet_id: int) -> dict:
		return requests.get(
			f"{self.api}/api/v2/sheets/{sheet_id}",
			headers=self.headers).json()

	def get_sheet_nodes(self, sheet_id: int) -> list:
		return requests.get(
			f"{self.api}/api/v2/sheets/{sheet_id}/nodes",
			headers=self.headers).json()

	def get_sheet_tabs(self, sheet_id: int) -> list:
		return requests.get(
			f"{self.api}/api/v2/sheets/{sheet_id}/tabs",
			headers=self.headers).json()

	def edit_game(
			self,
			game_id: int,
			name: str = None,
			description: str = None,
			rules: str = None,
			password: str = None,
			tags: list = None,
			max_players: int = None,
			expected_pace: int = None,
			beginner_friendly: bool = True,
			allow_spectators: bool = True,
			allow_public_chat: bool = True,
			allow_export: bool = False,
			is_hidden: bool = False) -> dict:
		data = {}
		if name:
			data["name"] = name
		if description:
			data["description"] = description
		if rules:
			data["rules"] = rules
		if password:
			data["password"] = password
		if tags:
			data["tags"] = tags
		if max_players:
			data["max_players"] = max_players
		if expected_pace:
			data["expected_pace"] = expected_pace
		if beginner_friendly:
			data["beginner_friendly"] = beginner_friendly
		if allow_spectators:
			data["allow_spectators"] = allow_spectators
		if allow_public_chat:
			data["allow_public_chat"] = allow_public_chat
		if allow_export:
			data["allow_export"] = allow_export
		if is_hidden:
			data["hidden"] = is_hidden
		return requests.patch(
			f"{self.api}/api/v2/games/{game_id}",
			data=data,
			headers=self.headers).json()

	def delete_game(self, game_id: int) -> int:
		return requests.delete(
			f"{self.api}/api/v2/games/{game_id}",
			headers=self.headers).status_code

	def get_message_groups(self) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{self.user_id}/message-groups",
			headers=self.headers).json()

	def search_message_group(self, query: str) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{self.user_id}/message-groups?search={query}",
			headers=self.headers).json()	

	def get_friend_requests(self) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{self.user_id}/friend-requests?receiver_id={self.user_id}",
			headers=self.headers).json()

	def get_user_profile(self, username: str) -> dict:
		return requests.get(
			f"{self.api}/api/v2/users/{username}",
			headers=self.headears).json()

	def change_email(self, email: str) -> dict:
		data = {
			"email": email
		}
		return requests.patch(
			f"{self.api}/api/v2/profiles/{self.user_id}",
			data=data,
			headers=self.headers).json()

	def delete_account(self) -> int:
		return requests.delete(
			f"{self.api}/api/v2/profiles/{self.user_id}",
			headers=self.headers).status_code

File: src/role_gate.js
class RoleGate {
	constructor() {
		this.api = "https://www.rolegate.com"
		this.headers = {
			"user-agent": "okhttp/3.14.9"
		}
	}

	async register(username, email, password) {
		const response = await fetch(
			`${this.api}/account/signup`, {
				method: "POST",
				body: JSON.stringify({
					username: username,
					password: password,
					password2: password,
					email: email
				}),
				headers: this.headers
			})
		return response.json()
	}

	async login(username, password) {
		const response = await fetch(
			`${this.api}/api/v2/api-token-auth`, {
				method: "POST",
				body: JSON.stringify({
					username: username,
					password: password
				}),
				headers: this.headers
			})
		const data = await response.json()
		if ("token" in data) {
			this.token = data.token
			this.headers["authorization"] = `Token ${this.token}`
			this.userId = await this.getCurrentUser().profile.id
		}
		return data
	}

	async getCurrentUser() {
		const response = await fetch(
			`${this.api}/api/v2/users/current`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getAvatars() {
		const response = await fetch(
			`${this.api}/api/v2/avatars`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getUserTransactions(userId, ordering = "date", pageSize = 50) {
		const response = await fetch(
			`${this.api}/api/v2/users/${userId}/coin-transactions?ordering=-${ordering}&page-size=${pageSize}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getUserReferringCount(userId) {
		const response = await fetch(
			`${this.api}/api/v2/users/${userId}/profiles-referring-count`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getUserGameAlerts(userId) {
		const response = await fetch(
			`${this.api}/api/v2/users/${userId}/game-alerts`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getUserBookmarks(userId, ordering = "date") {
		const response = await fetch(
			`${this.api}/api/v2/users/${userId}/games?ordering=-$${ordering}&bookmarked=True`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getFeaturedGames() {
		const response = await fetch(
			`${this.api}/api/v2/games/featured-games`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGameInfo(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGameCharacters(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/characters`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGamePreviewSelection(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/lines/preview-selection`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGameStates(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/states`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGameSheets(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/sheets`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGameTokens(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/tokens`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGameDiceSets(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/dice-sets`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async gerRecommendedTags() {
		const response = await fetch(
			`${this.api}/api/v2/recommended-tags`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getGames(page, isHidden = false) {
		const response = await fetch(
			`${this.api}/api/v2/games?hidden=${isHidden}&page=${page}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async joinGame(
			gameId,
			characterName,
			portraitAvatar = null,
			color = 1,
			isNpc = false) {
		let body = {
			color: color,
			background_color: 1,
			background_rank: "no",
			background_key: 0,
			"player-id": this.userId,
			npc: isNpc,
			"character-name": characterName
		}
		if (portraitAvatar) {
			body.portrait_avatar = portraitAvatar
		}
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}/characters`, {
				method: "POST",
				body: JSON.stringify(body),
				headers: this.headers
			})
		return response.json()
	}

	async leaveGame(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/users/${this.userId}/games/${gameId}/leave`, {
				method: "POST",
				headers: this.headers
			})
		return response.status_code
	}

	async createGame(
			name,
			description,
			rules,
			cover,
			sheetId,
			publicTemplateId,
			expectedPace = 0,
			isHidden = false,
			allowSpectators = true,
			allowPublicChat = true,
			maxPlayers = 4,
			password = null,
			tags = [],
			diceSets = [1]) {
		const response = await fetch(
			`${this.api}/api/v2/games`, {
				method: "POST",
				body: JSON.stringify({
					expected_pace: expectedPace,
					hidden: isHidden,
					allow_spectators: allowSpectators,
					allow_public_chat: allowPublicChat,
					max_players: maxPlayers,
					name: name,
					description: description,
					rules: rules,
					password: password,
					cover: cover,
					sheet: sheetId,
					tags: tags,
					dice_sets: diceSets,
					public_template: publicTemplateId
				}),
				headers: this.headers
			})
		return response.json()
	}

	async getBackgrounds(pageSize = 100) {
		const response = await fetch(
			`${this.api}/api/v2/stage-backgrounds?page-size=${pageSize}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getSheets(isPublic = true) {
		const response = await fetch(
			`${this.api}/api/v2/sheets?public=${isPublic}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async deleteGame(gameId) {
		const response = await fetch(
			`${this.api}/api/v2/games/${gameId}`, {
				method: "DELETE",
				headers: this.headers
			})
		return response.json()
	}
}

module.exports = {RoleGate}
